plutarch-extra/Plutarch/Extra/Map.hs

```diff
@@ -0,0 +1,324 @@
+ same set of keys. Requires a number of equality comparisons between keys
+ proportional to the length of the shorter argument.
+-}
+pkeysEqual ::
+  forall (k :: PType) (a :: PType) (b :: PType) (s :: S).
+  (PIsData k, PEq k) =>
+  Term s (PMap 'Sorted k a :--> PMap 'Sorted k b :--> PBool)
+pkeysEqual = phoistAcyclic $
+  plam $ \kvs kvs' ->
+    pmatch kvs $ \(PMap ell) ->
+      pmatch kvs' $ \(PMap ell') ->
+        go # ell # ell'
+  where
+    go ::
+      forall (s' :: S).
+      Term
+        s'
+        ( PBuiltinList (PBuiltinPair (PAsData k) (PAsData a))
+            :--> PBuiltinList (PBuiltinPair (PAsData k) (PAsData b))
+            :--> PBool
+        )
+    go = phoistAcyclic $
+      pfix #$ plam $ \self ell ell' ->
+        pmatch (PList.puncons # ell) $ \case
+          PNothing -> pmatch (PList.puncons # ell') $ \case
+            PNothing -> pcon PTrue -- no mismatches found
+            PJust _ -> pcon PFalse -- one argument too long
+          PJust kv -> pmatch (PList.puncons # ell') $ \case
+            PNothing -> pcon PFalse -- one argument too long
+            PJust kv' -> pmatch kv $ \(PPair h t) ->
+              pmatch kv' $ \(PPair h' t') ->
+                pif
+                  ((pkvPairKey # h) #== (pkvPairKey # h'))
+                  (self # t # t') -- continue
+                  (pcon PFalse) -- key mismatch
+
+-- | Get the key of a key-value pair.
+pkvPairKey ::
+  forall (k :: PType) (v :: PType) (s :: S).
+  (PIsData k) =>
+  Term s (PBuiltinPair (PAsData k) (PAsData v) :--> k)
+pkvPairKey = phoistAcyclic $ plam $ \kv -> pfromData (pfstBuiltin # kv)
+
+-- | Get the value of a key-value pair.
+pkvPairValue ::
+  forall (k :: PType) (v :: PType) (s :: S).
+  (PIsData v) =>
+  Term s (PBuiltinPair (PAsData k) (PAsData v) :--> v)
+pkvPairValue = phoistAcyclic $ plam $ \kv -> pfromData (psndBuiltin # kv)
+
+{- | Compare two key-value pairs by their keys. Gives 'PTrue' if the key of the
+ first argument pair is less than the key of the second argument pair.
+-}
+pkvPairLt ::
+  forall (k :: PType) (v :: PType) (s :: S).
+  (PIsData k, PPartialOrd k) =>
+  Term
+    s
+    ( PBuiltinPair (PAsData k) (PAsData v)
+        :--> PBuiltinPair (PAsData k) (PAsData v)
+        :--> PBool
+    )
+pkvPairLt = phoistAcyclic $
+  plam $ \kv kv' ->
+    (pkvPairKey # kv) #< (pkvPairKey # kv')
+
+-- | As 'plookup', but errors when the key is missing.
+ptryLookup ::
+  forall (k :: PType) (v :: PType) (keys :: KeyGuarantees) (s :: S).
+  (PIsData k, PIsData v) =>
+  Term s (k :--> PMap keys k v :--> v)
+ptryLookup = phoistAcyclic $
+  plam $ \k kvs ->
+    passertPJust # "plookupPartial: No value found for key."
+      # (plookup # k # kvs)
+
+{- | Given a 'Foldable' of key-value pairs, construct an unsorted 'PMap'.
+ Performs linearly with respect to its argument.
+
+ = Note
+
+ If there are duplicate keys in the input, the /last/ key will \'win\' in a
+ lookup.
+-}
+punsortedMapFromFoldable ::
```

Comment:
Isn't this basically an IsList instance? Why not just implement that?

Reply:
I don't think you can implement `toList` if we were to implement `IsList`

Reply:
at list without `plift`s which we don't want.

Reply:
I'm not sure what you mean.

Reply:
https://github.com/Liqwid-Labs/plutarch-quickcheck/blob/34f1157dd5d68d289beda59fc5c6f68da42d42c9/src/Plutarch/Test/QuickCheck/Instances.hs#L829

I can't think of a better way here. And I'm sure this isn't what we want.

Reply:
Yeah, the `toList` direction would be _horrid_.

Reply:
```haskell
instance IsList (Term s (PMap 'Unsorted k v)) where
  type Item (Term s (PMap 'Unsorted k v)) = [(Term s k, Term s v)]
  fromList = punsortedMapFromFoldable

instance IsList (Term s (PMap 'Sorted k v)) where
  type Item (Term s (PMap 'Unsorted k v)) = [(Term s k, Term s v)]
  fromList = psortedMapFromFoldable
```

Reply:
wait we can make `IsList` with no `toList`?